core/shared/src/main/scala/zio/internal/SingleThreadedRingBuffer.scala

```diff
@@ -48,7 +48,10 @@ private[zio] final class SingleThreadedRingBuffer[A <: AnyRef](capacity: Int) {
     var i                    = 0
     var result: List[AnyRef] = Nil
     while (i < array.length) {
-      result ::= array(i)
+      // https://github.com/zio/zio/issues/4384: We might concurrently drop the last
+      // element of the Ringbuffer and create the reversed list. Therefore we omit
+      // Null elements from the result.
+      if (Option(array(i)).isDefined) result ::= array(i)
```

Comment:
Please excuse an observation from a mere passer-by, but `Option(_)` allocates.
Wouldn't it be more frugal to just check `array(i) != null`? We're already using `var` and a `while` loop, so it wouldn't taint any pure FP here :D

Reply:
No need to excuse - well spotted. I have to get used to thinking in allocations on one hand and pure FP on the other. I have just updated my PR according to your suggestion.

Reply:
I have just see, the ci job failed. According to the test logs I don't think it is related to my change.

Reply:
other PRs are failing mysteriously in CI as well. Let me bring that up on Discord.